docs/source/rfc-configuration-overhaul.md

```diff
@@ -0,0 +1,893 @@
+configuration view, striking a balance between user-friendliness and software efficiency & correctness.
+
+#### Rationale
+
+1. **User Clarity and Software Efficiency:** Keeping a representation that directly mirrors the reference for users
+   (User Config) and a separate, robustly-typed structure for the software's operational use (Resolved Config) benefits
+   both users and developers.
+2. **Parse, Not Validate:** Transitioning from the User Config to the Resolved Config involves rigorous parsing and
+   validation. Upon the formation of the Resolved Config, it's inherently valid, minimising the likelihood of runtime
+   configuration errors.
+3. **Domain-Specific Considerations:** The transformation phase applies domain (in this case, Iroha) specific rules and
+   logic, ensuring the final Resolved Config is both syntactically and semantically apt.
+
+#### Research about Resolved Configuration
+
+##### Redundant Fields
+
+In the background about [redundant fields](#redundant-fields), it's mentioned that these fields were made available to
+end users primarily because the user config wasn't separated from the resolved config. However, in a two-stage
+configuration resolution, these redundant fields can be managed more efficiently.
+
+##### Root-Level `public_key` and `private_key`
+
+When users provide their configuration input, it should be possible for them to specify only the private key. From this,
+the public key can be derived. If both keys are provided, the system should derive the public key from the given private
+key for validation. A discrepancy between the derived public key and the provided one would lead to an error.
+
+##### Telemetry Configuration
+
+Considering the user config's structure:
+
+```rust
+struct Telemetry {
+    /// The node's name to be seen on the telemetry
+    name: Option<String>,
+    /// The url of the telemetry, e.g., ws://127.0.0.1:8001/submit
+    url: Option<Url>,
+    /// The minimum period of time in seconds to wait before reconnecting
+    min_retry_period: u64,
+    /// The maximum exponent of 2 that is used for increasing delay between reconnections
+    max_retry_delay_exponent: u8,
+    /// The filepath that to write dev-telemetry to
+    file: Option<std::path::PathBuf>,
+}
+```
+
+While this flat structure is convenient to represent the user configuration of telemetry (i.e. the content of
+`telemetry` section in the configuration file), there are some possible concerns:
+
+- If `name` is omitted, but some of the other fields are set, the system could warn that telemetry will not start unless
+  `name` is provided.
+- If the `name` is given, but neither `url` nor `file` are specified, the system should either issue a warning or
+  terminate, signaling that the telemetry configuration is incomplete.
+- If both the `url` and `file` fields are filled (indicating a potential conflict), a warning about the conflict should
```

Comment:
what kind of conflict might take place?

Reply:
I imply that `url` and `file` are mutually exclusive. I.e., Telemetry might have only one emit target.

Reply:
Added a note that here `url` and `file` are implied to be mutually exclusive for the purposes of this example.